### ISSUE TITLE
Self-encoded lengths that don't match

### ISSUE THREAD

**Issue by martinduke** [tlswg/tls13-spec#1219]
As far as I can tell, there is no requirement in RFC 8446 for receiving endpoints to verify that the self-encoded length on outer headers corresponds to the self-encoded lengths of component items. Mangled extensions that do this kind of thing are likely going to end up with decoding errors, but if the outer length is too long many implementations might survive, assuming they support the extension.

If it's not too onerous on implementations, it would be good to require a fatal alert if receivers detect this condition, to make behavior more predictable and ease debugging.

**Comment by davidben** [comment]
Which lengths are you referring to? I can't think of any length prefixes in RFC8446 that receivers are allowed to ignore. Are you saying that, if the length prefix on `opaque extension_data<0..2^16-1>` is longer than the structure inside it, receivers are allowed to ignore trailing data? I'd consider any implementation that allowed this non-compliant. The spec often contains text like:

>    The body of the "certificate_authorities" extension consists of a
>   CertificateAuthoritiesExtension structure.

It doesn't say the body has a prefix of a CertificateAuthoritiesExtension structure.

(And, of course, if the `extension_data` length prefix is too short, you won't be able to finish parsing the inner structure at all.)

**Comment by martinduke** [comment]
Obviously, bogus extension_data lengths are wrong, but there's no enforcement mechanism. If receivers MUST check that these match and abort if they don't, that would have the benefits described above. I think we're agreeing that ignoring trailing data would be wrong, but it's not actually written anywhere AFAICT.

Equally obviously, this is not the end of the world and TLS has apparently operated fine like this for a long time. But it's also the kind of problem that QUIC has spent a fair amount of energy to avoid.

**Comment by davidben** [comment]
I think it is already the case that receivers MUST check this. Receivers that don't are non-compliant. It's written somewhat implicitly by way of saying the body "consists of" such-and-such structure, but I think the most reasonable interpretation of this would be that you don't allow trailing data.

I'd support adding a sentence or two to the section 4.2 to clarify how extension definitions work in general. Though I've never heard of an implementation that misinterpreted this and did the wrong thing.

**Comment by richsalz** [comment]
Agree and @davidben, I don't know of any implementation that allows extending a field with arbitrary data and then ignoring it. They all do "I have 'n' bytes and am expecting a CertificateList" and if they finish with data left over it's a parse error.

**Comment by martinduke** [comment]
If the consensus is that this is a non-problem I am happy to have the editors close the issue.

**Comment by martinduke** [comment]
> I think it is already the case that receivers MUST check this. Receivers that don't are non-compliant. It's written somewhat implicitly by way of saying the body "consists of" such-and-such structure, but I think the most reasonable interpretation of this would be that you don't allow trailing data.

It's entirely possible that I missed a MUST in 8446 somewhere, but I couldn't find it.

**Comment by ekr** [comment]
I don't think a sentence to make this clear would be harmful.

**Comment by davidben** [comment]
How's https://github.com/tlswg/tls13-spec/pull/1220?
